src/infrastructure/retry.py

```
@@ -0,0 +1,168 @@
+        self.state = "closed"
+        self._lock = asyncio.Lock()
+
+    def __call__(self, func):
+        @wraps(func)
+        async def wrapper(*args, **kwargs):
+            async with self._lock:
+                if self.state == "open":
+                    if time.time() - self.last_failure_time > self.recovery_timeout:
+                        self.state = "half-open"
+                    else:
+                        raise DataSourceException("Circuit breaker is open")
```

Ruff (0.12.2)

Avoid specifying long messages outside the exception class

(TRY003)